apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -341,28 +374,82 @@ impl ApplyTo for PathSelection {
     }
 }
 
+impl ApplyTo for JSLiteral {
+    fn apply_to_path(
+        &self,
+        data: &JSON,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
+        errors: &mut IndexSet<ApplyToError>,
+    ) -> Option<JSON> {
+        match self {
+            Self::String(s) => Some(JSON::String(s.clone().into())),
+            Self::Number(n) => {
+                if let Ok(JSON::Number(n)) = serde_json_bytes::serde_json::from_str(n) {
+                    Some(JSON::Number(n))
+                } else {
+                    errors.insert(ApplyToError::new(
+                        format!("Invalid number {}", n).as_str(),
+                        input_path.to_vec(),
+                    ));
+                    None
+                }
+            }
+            Self::Bool(b) => Some(JSON::Bool(*b)),
+            Self::Null => Some(JSON::Null),
+            Self::Object(map) => {
+                let mut output = JSONMap::new();
+                for (key, value) in map {
+                    if let Some(value_json) = value.apply_to_path(data, vars, input_path, errors) {
+                        output.insert(key.clone(), value_json);
+                    }
+                }
+                Some(JSON::Object(output))
+            }
+            Self::Array(vec) => {
+                let mut output = vec![];
+                for value in vec {
+                    output.push(
+                        value
+                            .apply_to_path(data, vars, input_path, errors)
+                            .unwrap_or(JSON::Null),
+                    );
+                }
+                Some(JSON::Array(output))
+            }
+            Self::Path(path) => path.apply_to_path(data, vars, input_path, errors),
+        }
+    }
+}
+
 impl ApplyTo for SubSelection {
     fn apply_to_path(
         &self,
         data: &JSON,
-        vars: &IndexMap<String, JSON>,
-        input_path: &mut Vec<JSON>,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
         errors: &mut IndexSet<ApplyToError>,
     ) -> Option<JSON> {
         if let JSON::Array(array) = data {
             return self.apply_to_array(array, vars, input_path, errors);
         }
 
+        let vars: VarsWithPathsMap = {
+            let mut vars = vars.clone();
```

Comment:
This is going to get expensive when `$this` is huge. We either need the `Rc` improvement than we've been thinking about or a custom struct here that lets us temporarily override a variable.

Reply:
@nicholascioli were you looking into `Rc`ing vars?

Reply:
Yeah, but that work is on pause until this gets merged in since this PR changes much of the signatures and functionality of variables and the like.